netket/graph/symmetry.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from . import AbstractGraph
+from netket.utils.semigroup import SemiGroup
+from netket.utils.types import Array
+
+
+@dataclass(frozen=True)
+class SymmGroup(SemiGroup):
+    """
+    Collection of symmetry operations acting on the sites of a graph
+    (graph automorphisms).
+    """
+
+    graph: AbstractGraph
+    """Underlying graph"""
+
+    def __post_init__(self):
+        super().__post_init__()
+        myhash = hash((super().__hash__(), hash(self.graph)))
+        object.__setattr__(self, "_SymmGroup__hash", myhash)
+
+    def __matmul__(self, other):
+        if isinstance(other, SymmGroup) and self.graph != other.graph:
+            raise ValueError("Incompatible groups (underlying graph must be identical)")
+
+        return SymmGroup(super().__matmul__(other).elems, self.graph)
+
+    def to_array(self):
+        """
+        Convert the abstract group operations to an array of permutation indicies,
+        such that (for :code:`G = self`)::
+            V = np.array(G.graph.nodes())
+            assert np.all(G(V) == V[..., G.to_array()])
+        """
+        return self.__call__(np.arange(self.graph.n_nodes))
```

Comment:
do we need this? why can't we get rid of it?

Reply:
Hm, personally, I like having an explicit method since it is easier to discover than `asarray(group)`.